### PR TITLE
Remove false errors on pipes #refs 155

### PR DESCRIPTION
**Pull request by mk13**
Treat expressions with '|' as an 'asExpression' instead of as a 'bitwiseOrExpression'. Extend existing Dart parser to override the 'parseBitwiseOrExpression' to properly resolve piped expressions. 

Left side of pipe should be properly checked by the resolver while anything to the right of the pipe are ignored and will not raise any errors.

### CODE REVIEW

File: analyzer_plugin/lib/src/ng_expr_parser.dart [hunk 1]
```diff
@@ -0,0 +1,71 @@
+library angular2.src.analysis.analyzer_plugin.src.ng_expr_parser.dart;
```
Comment (MichaelRFairhurst) [inline analyzer_plugin/lib/src/ng_expr_parser.dart]: You can actually remove these, they don't do anything....

File: analyzer_plugin/lib/src/ng_expr_parser.dart [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import 'package:analyzer/src/generated/resolver.dart';
+import 'package:analyzer/dart/ast/standard_ast_factory.dart';
+import 'package:front_end/src/scanner/token.dart';
+
+class NgExprParser extends Parser {
+  NgExprParser(
+      Source source, AnalysisErrorListener errorListener, this.typeProvider)
+      : super(source, errorListener);
+
+  Token get _currentToken => super.currentToken;
```
Comment (MichaelRFairhurst) [inline analyzer_plugin/lib/src/ng_expr_parser.dart]: Hmm, is this necessary?

File: analyzer_plugin/lib/src/ng_expr_parser.dart [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  final TypeProvider typeProvider;
+
+  @override
```
Comment (MichaelRFairhurst) [inline analyzer_plugin/lib/src/ng_expr_parser.dart]: dangling annotation

File: analyzer_plugin/lib/src/ng_expr_parser.dart [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+   * Parse a bitwise or expression to be treated as a pipe.
+   * Return the resolved left-hand expression as a dynamic type.
+   *
+   *     bitwiseOrExpression ::=
+   *         bitwiseXorExpression ('|' pipeExpression)*
+   */
+  @override
+  Expression parseBitwiseOrExpression() {
+    Expression expression;
+    Token tempEndToken = null;
```
Comment (MichaelRFairhurst) [inline analyzer_plugin/lib/src/ng_expr_parser.dart]: maybe could use a name like 'beforePipeToken`

File: analyzer_plugin/lib/src/ng_expr_parser.dart [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    return expression;
+  }
+
+  /**
+   * Parse a bitwise or expression to be treated as a pipe.
+   * Return the resolved left-hand expression as a dynamic type.
+   *
+   *     pipeExpression ::= identifier[':' expression]*
+   */
+  void parsePipeExpression() {
```
Comment (MichaelRFairhurst) [inline analyzer_plugin/lib/src/ng_expr_parser.dart]: Very clean, very nice

File: analyzer_plugin/test/resolver_test.dart [hunk 6]
```diff
@@ -877,6 +877,75 @@ class GenericComponent<T extends String> {
...
+''');
+    _addHtmlSource(r"""
+<li *ngFor='let operator of (operators | pipe1)'>
+  {{operator.length}}
+</li>
+""");
+    _resolveSingleTemplate(dartSource);
+    errorListener.assertNoErrors();
+  }
+
```
Comment (MichaelRFairhurst) [inline analyzer_plugin/test/resolver_test.dart]: Some of the edge cases you found would be good to have as test cases, such as (a | number) + (b | number)

File: analyzer_plugin/lib/src/ng_expr_parser.dart (hunk 2 above)
Comment (mk13) [inline analyzer_plugin/lib/src/ng_expr_parser.dart]: It was flagging _currentToken as not found; it would probably work without it though. I can remove it though.

File: analyzer_plugin/test/resolver_test.dart (hunk 6 above)
Comment (mk13) [inline analyzer_plugin/test/resolver_test.dart]: Used my long gnarly parathensized expression chain instead.

**Comment by mk13** [comment]
Cool, changes have been made. And yea, I think it would be ready for next sdk release.

File: analyzer_plugin/lib/src/ng_expr_parser.dart (hunk 2 above)
Comment (MichaelRFairhurst) [inline analyzer_plugin/lib/src/ng_expr_parser.dart]: hmm, but shouldn't you be able to just do 'currentToken'?
